File: app/src/main/java/com/amarnehsoft/holyquran/network/NetworkModule.java
package com.amarnehsoft.holyquran.network;

import com.amarnehsoft.holyquran.base.App;
import com.squareup.moshi.Moshi;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
public class NetworkModule {

    @Provides
    @Singleton
    public OkHttpClient provideOkHttpClient(){
//        return new OkHttpClient.Builder().build();

        HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
        logging.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();
        httpClient.addInterceptor(logging);
        return httpClient.build();
    }

    @Provides
    @Singleton
    public Moshi provideMoshi(){
        return new Moshi.Builder().build();
    }

    @Provides
    @Singleton
    public Retrofit provideRetrofit(OkHttpClient httpClient, Moshi moshi, String baseUrl){
        return new Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(httpClient)
//                .addConverterFactory(MoshiConverterFactory.create(moshi))
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();
    }

    @Provides
    @Singleton
    public ApiService provideAppService(Retrofit retrofit){
        return retrofit.create(ApiService.class);
    }

    @Provides
    @Singleton
    public String provideBaseUrl(){
        return ServiceConstants.baseUrl;
    }

    @Provides
    @Singleton
    public NetworkUtils provideNetworkUtils(App app){
        return new NetworkUtils(app);
    }

}

File: app/src/main/java/com/amarnehsoft/holyquran/network/NetworkModule.kt
package com.amarnehsoft.holyquran.network

import com.amarnehsoft.holyquran.base.App
import com.amarnehsoft.holyquran.di.Quran
import com.amarnehsoft.holyquran.di.Tafseer
import com.amarnehsoft.holyquran.network.quran.QuranApi
import com.amarnehsoft.holyquran.network.tafseer.TafseerApi
import com.amarnehsoft.holyquran.utils.NetworkUtils

import javax.inject.Singleton

import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

@Module
class NetworkModule {

    @Provides
    @Singleton
    fun provideOkHttpClient(): OkHttpClient {
        val logging = HttpLoggingInterceptor()
        logging.level = HttpLoggingInterceptor.Level.BODY
        val httpClient = OkHttpClient.Builder()
        httpClient.addInterceptor(logging)
        return httpClient.connectTimeout(500, TimeUnit.SECONDS)
                .readTimeout(500, TimeUnit.SECONDS).build()
    }

    @Provides
    @Quran
    fun provideRetrofit(httpClient: OkHttpClient, @Quran baseUrl: String): Retrofit {
        return Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(httpClient)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
    }

    @Provides
    @Tafseer
    fun provideTafseerRetrofit(
            httpClient: OkHttpClient,
            @Tafseer baseUrl: String
    ): Retrofit {
        return Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(httpClient)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
    }

    @Provides
    @Singleton
    fun provideAppService(@Quran retrofit: Retrofit): QuranApi {
        return retrofit.create(QuranApi::class.java)
    }

    @Provides
    @Singleton
    fun provideTafseerApi(@Tafseer retrofit: Retrofit): TafseerApi {
        return retrofit.create(TafseerApi::class.java)
    }

    @Provides
    @Quran
    fun provideBaseUrl(): String {
        return ServiceConstants.quranBaseUrl
    }

    @Provides
    @Tafseer
    fun provideBaseUrlForTafseer(): String {
        return ServiceConstants.baseUrlForTafseer
    }

    @Provides
    @Singleton
    fun provideNetworkUtils(app: App): NetworkUtils {
        return NetworkUtils(app)
    }

    @Provides
    @Singleton
    fun provideRemoteDataSource(impl: RemoteDataSourceImpl): RemoteDataSource = impl
}
